README.md
[![badge](https://img.shields.io/twitter/follow/api_video?style=social)](https://twitter.com/intent/follow?screen_name=api_video) &nbsp; [![badge](https://img.shields.io/github/stars/apivideo/api.video-typescript-uploader?style=social)](https://github.com/apivideo/api.video-typescript-uploader) &nbsp; [![badge](https://img.shields.io/discourse/topics?server=https%3A%2F%2Fcommunity.api.video)](https://community.api.video)
![](https://github.com/apivideo/.github/blob/main/assets/apivideo_banner.png)
<h1 align="center">api.video typescript video uploader</h1>

![npm](https://img.shields.io/npm/v/@api.video/video-uploader) ![ts](https://badgen.net/badge/-/TypeScript/blue?icon=typescript&label)


[api.video](https://api.video) is the video infrastructure for product builders. Lightning fast video APIs for integrating, scaling, and managing on-demand & low latency live streaming features in your app.

# Table of contents

- [Table of contents](#table-of-contents)
- [Project description](#project-description)
- [Getting started](#getting-started)
  - [Installation](#installation)
    - [Installation method #1: requirejs](#installation-method-1-requirejs)
    - [Installation method #2: typescript](#installation-method-2-typescript)
    - [Simple include in a javascript project](#simple-include-in-a-javascript-project)
- [Documentation - Standard upload](#documentation---standard-upload)
  - [Instanciation](#instanciation)
    - [Options](#options)
      - [Using a delegated upload token (recommended):](#using-a-delegated-upload-token-recommended)
      - [Using an access token (discouraged):](#using-an-access-token-discouraged)
      - [Using an API key (**strongly** discouraged):](#using-an-api-key-strongly-discouraged)
      - [Common options](#common-options)
    - [Example](#example)
  - [Methods](#methods)
    - [`upload()`](#upload)
    - [`onProgress()`](#onprogress)
    - [`cancel()`](#cancel)
    - [`onPlayable()`](#onplayable)
- [Documentation - Progressive upload](#documentation---progressive-upload)
  - [Instanciation](#instanciation-1)
    - [Options](#options-1)
      - [Using a delegated upload token (recommended):](#using-a-delegated-upload-token-recommended-1)
      - [Using an access token (discouraged):](#using-an-access-token-discouraged-1)
      - [Common options](#common-options-1)
    - [Example](#example-1)
  - [Methods](#methods-1)
    - [`uploadPart(file: Blob)`](#uploadpartfile-blob)
    - [`uploadLastPart(file: Blob)`](#uploadlastpartfile-blob)
    - [`onProgress()`](#onprogress-1)
    - [`cancel()`](#cancel-1)
    - [`onPlayable()`](#onplayable-1)
- [Static wrapper](#static-wrapper)

# Project description

Typescript library to upload videos to api.video using delegated upload token (or usual access token) from the front-end. 

It allows you to upload videos in two ways:
- standard upload: to send a whole video file in one go
- progressive upload: to send a video file by chunks, without needing to know the final size of the video file

# Getting started

## Installation

### Installation method #1: requirejs

If you use requirejs you can add the library as a dependency to your project with 

```sh
$ npm install --save @api.video/video-uploader
```

You can then use the library in your script: 

```javascript
// standard upload:
var { VideoUploader } = require('@api.video/video-uploader');

var uploader = new VideoUploader({
    // ... (see bellow)
}); 

// progressive upload:
var { ProgressiveUploader } = require('@api.video/video-uploader');

var uploader = new ProgressiveUploader({
    // ... (see bellow)
}); 
```

### Installation method #2: typescript

If you use Typescript you can add the library as a dependency to your project with 

```sh
$ npm install --save @api.video/video-uploader
```

You can then use the library in your script: 

```typescript
// standard upload:
import { VideoUploader } from '@api.video/video-uploader'

const uploader = new VideoUploader({
    // ... (see bellow)
});

// progressive upload:
import { ProgressiveUploader } from '@api.video/video-uploader'

const uploader = new ProgressiveUploader({
    // ... (see bellow)
});
```


### Simple include in a javascript project

Include the library in your HTML file like so:

```html
<head>
    ...
    <script src="https://unpkg.com/@api.video/video-uploader" defer></script>
</head>
```

Then, once the `window.onload` event has been trigered, create your player using `new VideoUploader()`:
```html
...
<form>
    <input type="file" id="input" onchange="uploadFile(this.files)">
</form>
<script type="text/javascript">
    function uploadFile(files) {
        new VideoUploader({
            file: files[0],
            uploadToken: "YOUR_DELEGATED_TOKEN"
        }).upload();
    }
</script>
```

# Documentation - Standard upload

## Instanciation

### Options 

The upload library is instanciated using an `options` object. Options to provide depend on the way you want to authenticate to the API: either using a delegated upload token (recommanded), or using a usual access token. 

#### Using a delegated upload token (recommended):

Using delegated upload tokens for authentication is best options when uploading from the client side. To know more about delegated upload token, read the dedicated article on api.video's blog: [Delegated Uploads](https://api.video/blog/tutorials/delegated-uploads).


|                   Option name | Mandatory | Type   | Description             |
| ----------------------------: | --------- | ------ | ----------------------- |
|                   uploadToken | **yes**   | string | your upload token       |
|                       videoId | no        | string | id of an existing video |
| _common options (see bellow)_ |           |        |                         |

#### Using an access token (discouraged):

**Warning**: be aware that exposing your access token client-side can lead to huge security issues. Use this method only if you know what you're doing :).


|                   Option name | Mandatory | Type   | Description                                                                                                                                     |
| ----------------------------: | --------- | ------ | ----------------------------------------------------------------------------------------------------------------------------------------------- |
|                   accessToken | **yes**   | string | your access token                                                                                                                               |
|                  refreshToken | **no**    | string | your refresh token (please not that if you don't provide a refresh token, your upload may fails due to the access token lifetime of 60 minutes) |
|                       videoId | **yes**   | string | id of an existing video                                                                                                                         |
| _common options (see bellow)_ |           |        |                                                                                                                                                 |


#### Using an API key (**strongly** discouraged):

**Warning**: be aware that exposing your API key client-side can lead to huge security issues. Use this method only if you know what you're doing :).


|                   Option name | Mandatory | Type   | Description             |
| ----------------------------: | --------- | ------ | ----------------------- |
|                       API Key | **yes**   | string | your api.video API key  |
|                       videoId | **yes**   | string | id of an existing video |
| _common options (see bellow)_ |           |        |                         |


#### Common options


|   Option name    | Mandatory | Type                                                            | Description                                                                                                                              |
| ---------------: | --------- | --------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
|          file    | **yes**   | File                                                            | the file you want to upload                                                                                                              |
|     videoName    | no        | string                                                          | the name of your video (overrides the original file name for regular uploads, overrides the default "file" name for progressive uploads) |
|     chunkSize    | no        | number                                                          | number of bytes of each upload chunk (default: 50MB, min: 5MB, max: 128MB)                                                               |
|       apiHost    | no        | string                                                          | api.video host (default: ws.api.video)                                                                                                   |
|       retries    | no        | number                                                          | number of retries when an API call fails (default: 5)                                                                                    |
| retryStrategy    | no        | (retryCount: number, error: VideoUploadError) => number \| null | function that returns the number of ms to wait before retrying a failed upload. Returns null to stop retrying                            |
| maxVideoDuration | no        | number                                                          | maximum duration allowed for the file (in seconds)                                                                                       |


### Example

```javascript
    const uploader = new VideoUploader({
        file: files[0],
        uploadToken: "YOUR_DELEGATED_TOKEN",
        chunkSize: 1024*1024*10, // 10MB
        retries: 10,
    });
```

## Methods

### `upload()`

The upload() method starts the upload. It takes no parameter. It returns a Promise that resolves once the file is uploaded. If an API call fails more than the specified number of retries, then the promise is rejected.
On success, the promise embeds the `video` object returned by the API.
On fail, the promise embeds the status code & error message returned by the API.

**Example**

```javascript
    // ... uploader instanciation

    uploader.upload()
        .then((video) => console.log(video))
        .catch((error) => console.log(error.status, error.message));
```

### `onProgress()`

The onProgress() method let you defined an upload progress listener. It takes a callback function with one parameter: the onProgress events.
An onProgress event contains the following attributes:
- uploadedBytes (number): total number of bytes uploaded for this upload
- totalBytes (number): total size of the file
- chunksCount (number): number of upload chunks 
- chunksBytes (number): size of a chunk
- currentChunk (number): index of the chunk being uploaded
- currentChunkUploadedBytes (number): number of bytes uploaded for the current chunk

### `cancel()`

The cancel() method cancels the upload. It takes no parameter. 


**Example**

```javascript
    // ... uploader instanciation
    
    uploader.onProgress((event) => {
        console.log(`total number of bytes uploaded for this upload: ${event.uploadedBytes}.`);
        console.log(`total size of the file: ${event.totalBytes}.`);
        console.log(`number of upload chunks: ${event.chunksCount} .`);
        console.log(`size of a chunk: ${event.chunksBytes}.`);
        console.log(`index of the chunk being uploaded: ${event.currentChunk}.`);
        console.log(`number of bytes uploaded for the current chunk: ${event.currentChunkUploadedBytes}.`);
    });
```


### `onPlayable()`

The onPlayable() method let you defined a listener that will be called when the video is playable. It takes a callback function with one parameter: the `video` object returned by the API.

**Example**

```html
    <div id="player-container"></div>

    <script>
        // ... uploader instanciation
    
        uploader.onPlayable((video) => {
            // the video is playable, we can display the player
            document.getElementById('player-container').innerHTML = v.assets.iframe;
        });
    </script>
```

# Documentation - Progressive upload


## Instanciation

### Options 

The progressive upload object is instanciated using an `options` object. Options to provide depend on the way you want to authenticate to the API: either using a delegated upload token (recommanded), or using a usual access token. 

#### Using a delegated upload token (recommended):

Using delegated upload tokens for authentication is best options when uploading from the client side. To know more about delegated upload token, read the dedicated article on api.video's blog: [Delegated Uploads](https://api.video/blog/tutorials/delegated-uploads).


|                   Option name | Mandatory | Type   | Description             |
| ----------------------------: | --------- | ------ | ----------------------- |
|                   uploadToken | **yes**   | string | your upload token       |
|                       videoId | no        | string | id of an existing video |
| _common options (see bellow)_ |           |        |                         |

#### Using an access token (discouraged):

**Warning**: be aware that exposing your access token client-side can lead to huge security issues. Use this method only if you know what you're doing :).


|                   Option name | Mandatory | Type   | Description             |
| ----------------------------: | --------- | ------ | ----------------------- |
|                   accessToken | **yes**   | string | your access token       |
|                       videoId | **yes**   | string | id of an existing video |
| _common options (see bellow)_ |           |        |                         |


#### Common options


|       Option name           | Mandatory | Type                                                            | Description                                                                                                                                            |
| --------------------------: | --------- | --------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------ |
|           apiHost           | no        | string                                                          | api.video host (default: ws.api.video)                                                                                                                 |
|           retries           | no        | number                                                          | number of retries when an API call fails (default: 5)                                                                                                  |
|     retryStrategy           | no        | (retryCount: number, error: VideoUploadError) => number \| null | function that returns the number of ms to wait before retrying a failed upload. Returns null to stop retrying                                          |
| preventEmptyParts           | no        | boolean                                                         | if true, the upload will succeed even if an empty Blob is passed to uploadLastPart(). This may alter performances a bit in some cases (default: false) |
| mergeSmallPartsBeforeUpload | no        | boolean                                                         | if false, parts smaller than 5MB will not be merged before upload, resulting in an error (default: true)                                               |


### Example

```javascript
    const uploader = new ProgressiveUploader({
        uploadToken: "YOUR_DELEGATED_TOKEN",
        retries: 10,
    });
```

## Methods

### `uploadPart(file: Blob)`

The upload() method starts the upload. It takes no parameter. It returns a Promise that resolves once the file is uploaded. If an API call fails more than the specified number of retries, then the promise is rejected.
On success, the promise embeds the `video` object returned by the API.
On fail, the promise embeds the status code & error message returned by the API.

**Example**

```javascript
    // ... uploader instanciation

    uploader.uploadPart(blob)
        .catch((error) => console.log(error.status, error.message));
```

### `uploadLastPart(file: Blob)`

The upload() method starts the upload. It takes no parameter. It returns a Promise that resolves once the file is uploaded. If an API call fails more than the specified number of retries, then the promise is rejected.
On success, the promise embeds the `video` object returned by the API.
On fail, the promise embeds the status code & error message returned by the API.

**Example**

```javascript
    // ... uploader instanciation

    uploader.uploadLastPart(blob)
        .then((video) => console.log(video))
        .catch((error) => console.log(error.status, error.message));
```

### `onProgress()`

The onProgress() method let you defined an upload progress listener. It takes a callback function with one parameter: the onProgress events.
An onProgress event contains the following attributes:
- uploadedBytes (number): total number of bytes uploaded for this upload
- totalBytes (number): total size of the file
- part (number): index of the part being uploaded

**Example**

```javascript
    // ... uploader instanciation
    
    uploader.onProgress((event) => {
        console.log(`total number of bytes uploaded for this upload: ${event.uploadedBytes}.`);
        console.log(`total size of the file: ${event.totalBytes}.`);
        console.log(`current part: ${event.part}.`);
    });
```

### `cancel()`

The cancel() method cancels the upload. It takes no parameter. 

### `onPlayable()`

The onPlayable() method let you defined a listener that will be called when the video is playable. It takes a callback function with one parameter: the `video` object returned by the API.

**Example**

```html
    <div id="player-container"></div>

    <script>
        // ... uploader instanciation
    
        uploader.onPlayable((video) => {
            // the video is playable, we can display the player
            document.getElementById('player-container').innerHTML = v.assets.iframe;
        });
    </script>
```


# Static wrapper

For situations where managing object instances is impractical, consider using the [UploaderStaticWrapper](./doc/UploaderStaticWrapper.md) class, which offers static method equivalents for all functionalities.

CHANGELOG.md
# Changelog
All changes to this project will be documented in this file.

## [1.1.6] - 2023-10-31
- Add static wrapper

## [1.1.5] - 2023-10-26
- Add cancel() methods
- Add part number in progressive upload onProgress() 

## [1.1.4] - 2023-08-21
- Include `mp4` in VideoUploadResponse type

## [1.1.3] - 2023-01-23
- Add onPlayable() method to define a listener called when the video is playable
  
## [1.1.2] - 2023-01-19
- Video upload: add maxVideoDuration option.

## [1.1.1] - 2023-01-17
- Progressive upload: add mergeSmallPartsBeforeUpload option.

## [1.1.0] - 2022-07-06
- Video upload & Progressive upload: allow user to set a customized video name.

## [1.0.11] - 2022-07-06
- Add origin headers

## [1.0.10] - 2022-06-29
- Retry even if the server is not responding.
- Add possibility to define a custom retry policy.

## [1.0.9] - 2022-05-24
- Progressive upload: prevent last part to be empty

## [1.0.8] - 2022-04-27
- Create a AbstractUploader
- Add origin header
- Add the possibility to provide a refresh token

## [1.0.7] - 2022-04-26
- Don't retry on 401 errors
- Mutualize some code
  
## [1.0.6] - 2022-04-22
- Improve errors management

## [1.0.5] - 2022-04-21
- Fix date attributes types
- Add authentication using an API key

## [1.0.4] - 2022-03-23
- Export `VideoUploadResponse` type

## [1.0.3] - 2022-01-25
- Fix typo in `VideoUploadResponse`type: `hsl` instead of `hls`
  
## [1.0.2] - 2021-11-24
- Fix: prevent concurrent requests
  
## [1.0.1] - 2021-11-23
- Add missing return types in upload methods
  
## [1.0.0] - 2021-11-16
- Bump dependancies
  
## [0.0.6] - 2021-11-10
- Add progressive upload feature

## [0.0.5] - 2021-08-12
- Fix chunk size (default 50MB, min 5MB, max 128MB)
- Fix upload progress listener


src/abstract-uploader.ts
export const MIN_CHUNK_SIZE = 1024 * 1024 * 5; // 5mb
export const DEFAULT_CHUNK_SIZE = 1024 * 1024 * 50; // 50mb
export const MAX_CHUNK_SIZE = 1024 * 1024 * 128; // 128mb
export const DEFAULT_RETRIES = 6;
export const DEFAULT_API_HOST = "ws.api.video";

export declare type VideoUploadResponse = {
    readonly videoId: string;
    readonly title?: string;
    readonly description?: string;
    readonly _public?: boolean;
    readonly panoramic?: boolean;
    readonly mp4Support?: boolean;
    readonly publishedAt?: Date;
    readonly createdAt?: Date;
    readonly updatedAt?: Date;
    readonly tags?: string[];
    readonly metadata?: {
        readonly key?: string;
        readonly value?: string;
    }[];
    readonly source?: {
        readonly type?: string;
        readonly uri?: string;
    };
    readonly assets?: {
        readonly iframe?: string;
        readonly player?: string;
        readonly hls?: string;
        readonly thumbnail?: string;
        readonly mp4?: string;
    };
};

type RetryStrategy = (
    retryCount: number,
    error: VideoUploadError,
) => number | null;

interface Origin {
    name: string;
    version: string;
}

export interface CommonOptions {
    apiHost?: string;
    retries?: number;
    videoName?: string;
    retryStrategy?: RetryStrategy;
    origin?: {
        application?: Origin;
        sdk?: Origin;
    };
}

export interface WithUploadToken {
    uploadToken: string;
    videoId?: string;
}

export interface WithAccessToken {
    accessToken: string;
    refreshToken?: string;
    videoId: string;
}

export interface WithApiKey {
    apiKey: string;
    videoId: string;
}

export type VideoUploadError = {
    status?: number;
    type?: string;
    title?: string;
    reason?: string;
    raw: string;
};

type HXRRequestParams = {
    parts?: {
        currentPart: number;
        totalParts: number | "*";
    };
    onProgress?: (e: ProgressEvent) => void;
    body: Document | XMLHttpRequestBodyInit | null;
};

export interface CancelableOperation<T> {
    cancel: () => void;
    result: Promise<T>;
}

let PACKAGE_VERSION = "";
try {
    // @ts-ignore
    PACKAGE_VERSION = __PACKAGE_VERSION__ || "";
} catch (e) {
    // ignore
}

export const DEFAULT_RETRY_STRATEGY = (maxRetries: number) => {
    return (retryCount: number, error: VideoUploadError) => {
        if (
            (error.status && error.status >= 400 && error.status < 500) ||
            retryCount >= maxRetries
        ) {
            return null;
        }
        return Math.floor(200 + 2000 * retryCount * (retryCount + 1));
    };
};

export abstract class AbstractUploader<T> {
    protected uploadEndpoint: string;
    protected videoId?: string;
    protected retries: number;
    protected headers: { [name: string]: string } = {};
    protected onProgressCallbacks: ((e: T) => void)[] = [];
    protected onPlayableCallbacks: ((e: VideoUploadResponse) => void)[] = [];
    protected refreshToken?: string;
    protected apiHost: string;
    protected retryStrategy: RetryStrategy;
    protected abortControllers: { [id: string]: AbortController } = {};

    constructor(
        options: CommonOptions & (WithAccessToken | WithUploadToken | WithApiKey),
    ) {
        this.apiHost = options.apiHost || DEFAULT_API_HOST;

        if (options.hasOwnProperty("uploadToken")) {
            const optionsWithUploadToken = options as WithUploadToken;
            if (optionsWithUploadToken.videoId) {
                this.videoId = optionsWithUploadToken.videoId;
            }
            this.uploadEndpoint = `https://${this.apiHost}/upload?token=${optionsWithUploadToken.uploadToken}`;
        } else if (options.hasOwnProperty("accessToken")) {
            const optionsWithAccessToken = options as WithAccessToken;
            if (!optionsWithAccessToken.videoId) {
                throw new Error("'videoId' is missing");
            }
            this.refreshToken = optionsWithAccessToken.refreshToken;
            this.uploadEndpoint = `https://${this.apiHost}/videos/${optionsWithAccessToken.videoId}/source`;
            this.headers.Authorization = `Bearer ${optionsWithAccessToken.accessToken}`;
        } else if (options.hasOwnProperty("apiKey")) {
            const optionsWithApiKey = options as WithApiKey;
            if (!optionsWithApiKey.videoId) {
                throw new Error("'videoId' is missing");
            }
            this.uploadEndpoint = `https://${this.apiHost}/videos/${optionsWithApiKey.videoId}/source`;
            this.headers.Authorization = `Basic ${btoa(
                optionsWithApiKey.apiKey + ":",
            )}`;
        } else {
            throw new Error(
                `You must provide either an accessToken, an uploadToken or an API key`,
            );
        }
        this.headers["AV-Origin-Client"] = "typescript-uploader:" + PACKAGE_VERSION;
        this.retries = options.retries || DEFAULT_RETRIES;
        this.retryStrategy =
            options.retryStrategy || DEFAULT_RETRY_STRATEGY(this.retries);

        if (options.origin) {
            if (options.origin.application) {
                AbstractUploader.validateOrigin(
                    "application",
                    options.origin.application,
                );
                this.headers[
                    "AV-Origin-App"
                ] = `${options.origin.application.name}:${options.origin.application.version}`;
            }
            if (options.origin.sdk) {
                AbstractUploader.validateOrigin("sdk", options.origin.sdk);
                this.headers[
                    "AV-Origin-Sdk"
                ] = `${options.origin.sdk.name}:${options.origin.sdk.version}`;
            }
        }
    }

    public onProgress(cb: (e: T) => void) {
        this.onProgressCallbacks.push(cb);
    }

    public onPlayable(cb: (e: VideoUploadResponse) => void) {
        this.onPlayableCallbacks.push(cb);
    }

    protected async waitForPlayable(video: VideoUploadResponse) {
        const hls = video.assets?.hls;

        while (true) {
            await this.sleep(500);

            const hlsRes = await fetch(hls!);

            if (hlsRes.status === 202) {
                continue;
            }

            if ((await hlsRes.text()).length === 0) {
                continue;
            }

            break;
        }

        this.onPlayableCallbacks.forEach((cb) => cb(video));
    }

    protected parseErrorResponse(xhr: XMLHttpRequest): VideoUploadError {
        try {
            const parsedResponse = JSON.parse(xhr.response);

            return {
                status: xhr.status,
                raw: xhr.response,
                ...parsedResponse,
            };
        } catch (e) {
            // empty
        }

        return {
            status: xhr.status,
            raw: xhr.response,
            reason: "UNKWOWN",
        };
    }

    protected apiResponseToVideoUploadResponse(
        response: any,
    ): VideoUploadResponse {
        const res = {
            ...response,
            _public: response.public,
            publishedAt: response.publishedAt
                ? new Date(response.publishedAt)
                : undefined,
            createdAt: response.createdAt ? new Date(response.createdAt) : undefined,
            updatedAt: response.updatedAt ? new Date(response.updatedAt) : undefined,
        };
        delete res.public;
        return res;
    }

    protected sleep(duration: number): Promise<void> {
        return new Promise((resolve, reject) => {
            setTimeout(() => resolve(), duration);
        });
    }

    protected xhrWithRetrier(
        params: HXRRequestParams,
    ): CancelableOperation<VideoUploadResponse> {
        return this.withRetrier((abortController: AbortController) =>
            this.createXhrPromise(params, abortController),
        );
    }

    protected createFormData(
        file: Blob,
        fileName: string,
        startByte?: number,
        endByte?: number,
    ): FormData {
        const chunk = startByte || endByte ? file.slice(startByte, endByte) : file;
        const chunkForm = new FormData();
        if (this.videoId) {
            chunkForm.append("videoId", this.videoId);
        }
        chunkForm.append("file", chunk, fileName);
        return chunkForm;
    }

    public doRefreshToken(): Promise<void> {
        return new Promise((resolve, reject) => {
            const xhr = new window.XMLHttpRequest();
            xhr.open("POST", `https://${this.apiHost}/auth/refresh`);
            for (const headerName of Object.keys(this.headers)) {
                if (headerName !== "Authorization")
                    xhr.setRequestHeader(headerName, this.headers[headerName]);
            }
            xhr.onreadystatechange = (_) => {
                if (xhr.readyState === 4 && xhr.status >= 400) {
                    reject(this.parseErrorResponse(xhr));
                }
            };
            xhr.onload = (_) => {
                const response = JSON.parse(xhr.response);
                if (response.refresh_token && response.access_token) {
                    this.headers.Authorization = `Bearer ${response.access_token}`;
                    this.refreshToken = response.refresh_token;
                    resolve();
                    return;
                }
                reject(this.parseErrorResponse(xhr));
            };
            xhr.send(
                JSON.stringify({
                    refreshToken: this.refreshToken,
                }),
            );
        });
    }
    private createXhrPromise(
        params: HXRRequestParams,
        abortController: AbortController,
    ): Promise<VideoUploadResponse> {
        return new Promise((resolve, reject) => {
            const xhr = new window.XMLHttpRequest();
            xhr.open("POST", `${this.uploadEndpoint}`, true);
            abortController.signal.addEventListener("abort", () => {
                xhr.abort();
                reject({
                    status: undefined,
                    raw: undefined,
                    reason: "ABORTED",
                });
            });
            if (params.parts) {
                xhr.setRequestHeader(
                    "Content-Range",
                    `part ${params.parts.currentPart}/${params.parts.totalParts}`,
                );
            }
            for (const headerName of Object.keys(this.headers)) {
                xhr.setRequestHeader(headerName, this.headers[headerName]);
            }
            if (params.onProgress) {
                xhr.upload.onprogress = (e) => params.onProgress!(e);
            }
            xhr.onreadystatechange = (_) => {
                if (xhr.readyState === 4) {
                    // DONE
                    if (xhr.status === 401 && this.refreshToken) {
                        return this.doRefreshToken()
                            .then(() => this.createXhrPromise(params, abortController))
                            .then((res) => resolve(res))
                            .catch((e) => reject(e));
                    } else if (xhr.status >= 400) {
                        reject(this.parseErrorResponse(xhr));
                        return;
                    }
                }
            };
            xhr.onerror = (e) => {
                reject({
                    status: undefined,
                    raw: undefined,
                    reason: "NETWORK_ERROR",
                });
            };
            xhr.ontimeout = (e) => {
                reject({
                    status: undefined,
                    raw: undefined,
                    reason: "NETWORK_TIMEOUT",
                });
            };
            xhr.onload = (_) => {
                if (xhr.status < 400) {
                    resolve(
                        this.apiResponseToVideoUploadResponse(JSON.parse(xhr.response)),
                    );
                }
            };
            xhr.send(params.body);
        });
    }

    private withRetrier(
        fn: (abortController: AbortController) => Promise<VideoUploadResponse>,
    ): CancelableOperation<VideoUploadResponse> {
        // generate a unique random id for this upload
        const id =
            Math.random().toString(36).substring(2, 15) +
            Math.random().toString(36).substring(2, 15);
        const abortController = new AbortController();
        this.abortControllers[id] = abortController;

        const promise = new Promise<VideoUploadResponse>(
            async (resolve, reject) => {
                let retriesCount = 0;
                while (true) {
                    try {
                        const res = await fn(abortController);
                        resolve(res);
                        return;
                    } catch (e: any) {
                        if (e.reason === "ABORTED") {
                            reject(e);
                            return;
                        }
                        const retryDelay = this.retryStrategy(retriesCount, e);
                        if (retryDelay === null) {
                            reject(e);
                            return;
                        }
                        console.log(
                            `video upload: ${e.reason || "ERROR"
                            }, will be retried in ${retryDelay} ms`,
                        );
                        await this.sleep(retryDelay);
                        retriesCount++;
                    }
                }
            },
        );

        return {
            cancel: () => {
                this.abortControllers[id].abort();
                delete this.abortControllers[id];
            },
            result: promise,
        };
    }

    private static validateOrigin(type: string, origin: Origin) {
        if (!origin.name) {
            throw new Error(`${type} name is required`);
        }
        if (!origin.version) {
            throw new Error(`${type} version is required`);
        }
        if (!/^[\w-]{1,50}$/.test(origin.name)) {
            throw new Error(
                `Invalid ${type} name value. Allowed characters: A-Z, a-z, 0-9, '-', '_'. Max length: 50.`,
            );
        }
        if (!/^\d{1,3}(\.\d{1,3}(\.\d{1,3})?)?$/.test(origin.version)) {
            throw new Error(
                `Invalid ${type} version value. The version should match the xxx[.yyy][.zzz] pattern.`,
            );
        }
    }
}
